File: algorithms/Last Stone Weight II/Last Stone Weight II.cpp
class Solution {
public:
    int lastStoneWeightII(vector<int>& stones) {
        int n=stones.size();
        int sum=0;
        for(int i=0;i<n;i++){
            sum+=stones[i];
        }
        
        int dp[n+1][sum+1];
        for(int i=0;i<n+1;i++){
            for(int j=0;j<sum+1;j++){
                if(i==0)
                    dp[i][j]=0;
                if(j==0)
                    dp[i][j]=1;
            }
        }
        for(int i=1;i<n+1;i++){
            for(int j=1;j<sum+1;j++){
                if(stones[i-1]<=j){
                    dp[i][j]=dp[i-1][j-stones[i-1]] || dp[i-1][j];
                }
                else
                    dp[i][j]=dp[i-1][j];
            }
        }
        vector<int>v;
        for(int i=0;i<=sum/2;i++){
            if(dp[n][i]==1)
                v.push_back(i);
        }
        int mn=INT_MAX;
        for(int i=0;i<v.size();i++){
            mn=min(mn,sum-2*v[i]);
        }
        return mn;
    }
};

File: scripts/algorithms/L/Last Stone Weight II/Last Stone Weight II.js
// Runtime: 133 ms (Top 21.62%) | Memory: 42.3 MB (Top 86.49%)
var lastStoneWeightII = function(stones) {
    var sum = stones.reduce((a, b) => a + b);
    var target = Math.ceil(sum / 2);
    var dp = new Array(target + 1).fill(0);
    for(var i = 0; i < stones.length; i++) {
        for(var j = target; j >= stones[i]; j--) {
            dp[j] = Math.max(dp[j], dp[j - stones[i]] + stones[i]);
        }
    }
    return Math.abs(2 * dp[target] - sum);
};